File: algorithms/graph/spfa/csharp/Spfa.cs
using System;
using System.Collections.Generic;

public class Spfa
{
    public static int Solve(int[] arr)
    {
        int n = arr[0];
        int m = arr[1];
        int src = arr[2];
        var adj = new List<(int v, int w)>[n];
        for (int i = 0; i < n; i++) adj[i] = new List<(int, int)>();
        for (int i = 0; i < m; i++)
        {
            int u = arr[3 + 3 * i];
            int v = arr[3 + 3 * i + 1];
            int w = arr[3 + 3 * i + 2];
            adj[u].Add((v, w));
        }

        int INF = int.MaxValue / 2;
        int[] dist = new int[n];
        for (int i = 0; i < n; i++) dist[i] = INF;
        dist[src] = 0;
        bool[] inQueue = new bool[n];
        var queue = new Queue<int>();
        queue.Enqueue(src);
        inQueue[src] = true;

        while (queue.Count > 0)
        {
            int u = queue.Dequeue();
            inQueue[u] = false;
            foreach (var (v, w) in adj[u])
            {
                if (dist[u] + w < dist[v])
                {
                    dist[v] = dist[u] + w;
                    if (!inQueue[v])
                    {
                        queue.Enqueue(v);
                        inQueue[v] = true;
                    }
                }
            }
        }

        return dist[n - 1] == INF ? -1 : dist[n - 1];
    }
}

File: algorithms/graph/spfa/scala/Spfa.scala
object Spfa {

  def spfa(arr: Array[Int]): Int = {
    val n = arr(0)
    val m = arr(1)
    val src = arr(2)
    val adj = Array.fill(n)(scala.collection.mutable.ListBuffer[(Int, Int)]())
    for (i <- 0 until m) {
      val u = arr(3 + 3 * i)
      val v = arr(3 + 3 * i + 1)
      val w = arr(3 + 3 * i + 2)
      adj(u) += ((v, w))
    }

    val INF = Int.MaxValue / 2
    val dist = Array.fill(n)(INF)
    dist(src) = 0
    val inQueue = Array.fill(n)(false)
    val queue = scala.collection.mutable.Queue[Int]()
    queue.enqueue(src)
    inQueue(src) = true

    while (queue.nonEmpty) {
      val u = queue.dequeue()
      inQueue(u) = false
      for ((v, w) <- adj(u)) {
        if (dist(u) + w < dist(v)) {
          dist(v) = dist(u) + w
          if (!inQueue(v)) {
            queue.enqueue(v)
            inQueue(v) = true
          }
        }
      }
    }

    if (dist(n - 1) == INF) -1 else dist(n - 1)
  }
}
